Rotate CSGObjects and coordinate with SetGoniometer and CopySample

DanielMurphy22 opened the pull request:
The following shapes can be rotated, and the finite ones can be plotted:
Cuboid
Cylinder
HollowCylinder
InfinitePlane
InfiniteCylinder
Hexahedron
TaperedGuide
SliceofCylinderRing

The rotation can be manually set by the user using the rotate tag
Further rotation is added using a goniometer tag, which works with SetGoniometer

CopySample can now copy a shape which can be plotted, but maintains its own
Goniometer angle

**Description of work.**

**To test:**

1) Run the below script. To rotate the cuboid around the x,y,z axes (in that order) change the values in the rotate tag in `cuboid_xml`. Check this works for [other shapes too](https://docs.mantidproject.org/nightly/concepts/HowToDefineGeometricShape.html#howtodefinegeometricshape), and that infinite cylinders and infinite cuboids cannot be plotted!

<details><summary>Rotate CSGObjects Script</summary>
<p>

``` python
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
from mpl_toolkits.mplot3d.art3d import Poly3DCollection
from mantid.api import AnalysisDataService as ADS

ws = CreateSampleWorkspace()

cuboid_xml = " \
<cuboid id='some-cuboid'> \
  <height val='20.0'  /> \
  <width val='50.0' />  \
  <depth  val='20.0' />  \
  <centre x='100.0' y='100.0' z='100.0'  />  \
  <rotate x='0' y='0' z='0' /> \
</cuboid>  \
<algebra val='some-cuboid' /> \
"

# cylinder_xml = ' \
# <cylinder id="stick"> \
#   <centre-of-bottom-base x="-1.0" y="0.0" z="0.0" /> \
#   <axis x="1.0" y="0.0" z="0.0" /> \
#   <radius val="0.2" /> \
#   <height val="2.0" /> \
#   <rotate x=\"0\" y=\"0\" z=\"0\" /> \
# </cylinder> \
# <algebra val=\'some-cylinder\' /> \
# '
# # 
# merge_xml = ' \
# <cylinder id="stick"> \
#   <centre-of-bottom-base x="-0.5" y="0.0" z="0.0" /> \
#   <axis x="1.0" y="0.0" z="0.0" />  \
#   <radius val="0.05" /> \
#   <height val="1.0" /> \
# </cylinder> \
#  \
# <sphere id="some-sphere"> \
#   <centre x="0.0"  y="0.0" z="0.0" /> \
#   <radius val="0.5" /> \
# </sphere> \
#  \
# <algebra val="some-sphere (# stick)" /> \
# '
# # 
# slice = ' \
# <slice-of-cylinder-ring id="A"> \
#   <inner-radius val="0.0" /> \
#   <outer-radius val="0.0646" /> \
#   <depth val="0.01" /> \
#   <arc val="90.0" />  \
#   <rotate x="90.0" y="45.0" z="180.0" /> \
# </slice-of-cylinder-ring> \
# '
# inf_cylinder = ' \
# <infinite-cylinder id="A" > \
#   <centre x="0.0" y="0.2" z="0" /> \
#   <axis x="0.0" y="0.2" z="0" /> \
#   <radius val="1" /> \
# </infinite-cylinder> \
# '
# tapered-guide = ' \
# <tapered-guide id="A Guide"> \
#     <aperture-start height="2.0" width="2.0" /> \
#     <length val="3.0" /> \
#     <aperture-end height="4.0" width="4.0" /> \
#     <centre x="0.0" y="5.0" z="10.0" /> \ <!-- Optional.  Defaults to (0, 0 ,0) --> \
#     <axis x="0.0" y="0.0" z="1.0" /> \ <!-- Optional.  Defaults to (0, 0 ,1) --> \
# </tapered-guide> \
# '

# Set sample geometry of workspace to this CSG object
SetSample(ws, Geometry={'Shape': 'CSG', 'Value': cuboid_xml})

sample = ws.sample()
shape = sample.getShape()
mesh = shape.getMesh()

print(mesh)

facecolors = ['b','royalblue','red','firebrick', 'orange', 'gold','green', 'darkgreen','grey','black', 'purple', 'mediumorchid']

mesh_polygon_a = Poly3DCollection(mesh, facecolors = facecolors, linewidths=0.1)
'''Problem based on order of being added!'''

fig, axes = plt.subplots(subplot_kw={'projection':'mantid3d'})
axes.add_collection3d(mesh_polygon_a)

# Auto scale to the mesh size
axes_lims = mesh.flatten()
axes.auto_scale_xyz(axes_lims, axes_lims, axes_lims)

axes.set_title('Sample Shape: Tube')
axes.set_xlabel('X / m')
axes.set_ylabel('Y / m')
axes.set_zlabel('Z / m')

plt.show()
```
</p>
</details>

Fixes #31449

2) Here is a script to test that CopySample now works with these new rotations by the goniometer. The first setGoniometer on ws does no rotation, but SetGoniometer on ws1 rotates about the y-axis clockwise by 90 degrees. CopySample should be able to copy over the shape (cuboid), but correctly use the goniometer angles for ws1.

<details><summary>SetGoniometer and CopySample for CSGObjects Script</summary>
<p>

``` python
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
from mpl_toolkits.mplot3d.art3d import Poly3DCollection
from mantid.api import AnalysisDataService as ADS


# A fake host workspace, replace this with your real one.
ws = CreateSampleWorkspace()
SetGoniometer(ws, Axis0="0,0,1,0,1")

# Specify a Sphere geometry using CSG

cuboid_alt_xml = " \
<cuboid id='some-cuboid'> \
  <height val='20.0'  /> \
  <width val='50.0' />  \
  <depth  val='20.0' />  \
  <centre x='100.0' y='100.0' z='100.0'  />  \
  <rotate x='0' y='0' z='0' /> \
</cuboid>  \
<algebra val='some-cuboid' /> \
"

# Set sample geometry of workspace to this CSG object Sphere
SetSample(ws, Geometry={'Shape': 'CSG', 'Value': cuboid_alt_xml})

ws1 = CreateSampleWorkspace()
SetGoniometer(ws1, Axis0="90,0,1,0,1")
CopySample(ws,ws1,CopyEnvironment=False, CopyMaterial=False,CopyShape=True)
 
for i,val in enumerate((ws,ws1)): 
    sample = val.sample()
    shape = sample.getShape()
    mesh = shape.getMesh()

    facecolors = ['royalblue','b','red','firebrick', 'gold', 'orange','green', 'darkgreen','grey','black', 'purple', 'mediumorchid']

    mesh_polygon = Poly3DCollection(mesh, facecolors = facecolors, linewidths=0.1)

    fig, axes = plt.subplots(subplot_kw={'projection':'mantid3d'})
    axes.add_collection3d(mesh_polygon)

    # Auto scale to the mesh size
    axes_lims = mesh.flatten()
    axes.auto_scale_xyz(axes_lims, axes_lims, axes_lims)

    axes.set_title('Sample Shape: Cuboid ws{}'.format(i))
    axes.set_xlabel('X / m')
    axes.set_ylabel('Y / m')
    axes.set_zlabel('Z / m')

    plt.show()
```
</p>
</details>

3) Here is a script to check that the CopySample implementation also works for MeshObjects now too. You will find tube.stl on the systemtest data.

<details><summary>CopySample for MeshObjects Script</summary>
<p>

``` python
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

from mpl_toolkits.mplot3d.art3d import Poly3DCollection
from mantid.api import AnalysisDataService as ADS

# load sample shape mesh file for a workspace
ws = CreateSampleWorkspace()
# alternatively: ws = Load('filepath') or ws = ADS.retrieve('ws')
SetGoniometer(ws, Axis0="0,0,1,0,1")
ws = LoadSampleShape(ws, "tube.stl")

ws1 = CreateSampleWorkspace()
SetGoniometer(ws1, Axis0="90,0,1,0,1")
CopySample(ws,ws1,CopyEnvironment=False, CopyMaterial=False,CopyShape=True)

for i,val in enumerate((ws,ws1)): 
    sample = val.sample()
    shape = sample.getShape()
    mesh = shape.getMesh()

    facecolors = ['royalblue','b','red','firebrick', 'gold', 'orange','green', 'darkgreen','grey','black', 'purple', 'mediumorchid']

    mesh_polygon = Poly3DCollection(mesh, facecolors = facecolors, linewidths=0.1)

    fig, axes = plt.subplots(subplot_kw={'projection':'mantid3d'})
    axes.add_collection3d(mesh_polygon)

    # Auto scale to the mesh size
    axes_lims = mesh.flatten()
    axes.auto_scale_xyz(axes_lims, axes_lims, axes_lims)

    axes.set_title('Sample Shape: Cuboid ws{}'.format(i))
    axes.set_xlabel('X / m')
    axes.set_ylabel('Y / m')
    axes.set_zlabel('Z / m')

    plt.show()
```
</p>
</details>

DanielMurphy22 committed:
Rotate CSGObjects and coordinate with SetGoniometer and CopySample

The following shapes can be rotated, and the finite ones can be plotted:
Cuboid
Cylinder
HollowCylinder
InfinitePlane
InfiniteCylinder
Hexahedron
TaperedGuide
SliceofCylinderRing

The rotation can be manually set by the user using the rotation tag
Further rotation is added using a goniometer tag, which works with SetGoniometer

CopySample can now copy a shape which can be plotted, but maintains its own
Goniometer angle

DannyHindson commented on Framework/Algorithms/src/CopySample.cpp:
```diff
@@ -150,6 +158,52 @@ void CopySample::copyParameters(Sample &from, Sample &to, bool nameFlag, bool ma
+    if (auto csgObj = std::dynamic_pointer_cast<Geometry::CSGObject>(rhsObject)) {
+      // Rotate CSGObject by goniometer by editing XML, if possible for that shape
+
+      std::string xml = rhsObject->getShapeXML();
```
If you did csgObj->getShapeXML() here would that remove the need to define the getShapeXML method on all of the IObject classes?

DannyHindson commented on Framework/Algorithms/src/CopySample.cpp:
```diff
@@ -150,6 +158,52 @@ void CopySample::copyParameters(Sample &from, Sample &to, bool nameFlag, bool ma
+        goniometerRotation += matrixElementNames[index] + " = '" + std::to_string(rotationMatrix[index]) + "' ";
+      }
+      goniometerRotation += "/>";
+      xml.insert(gonioPlace, goniometerRotation);
```
The 10-20 lines here is also in the SetSample.cpp unit. Perhaps this could be moved into the ShapeFactor as a new method? The method could take the rotation matrix and the shape xml as parameters perhaps and insert the goniometer tag

DannyHindson reviewed (commented):
I've had an initial look through and have added a couple of comments

DannyHindson commented:
@peterfpeterson, @gvardany - we've created this PR to rotate the sample whenever a sample shape is added to a workspace that has a goniometer rotation (eg using SetSample or CopySample). This will happen for CSG or mesh sample shapes. The goniometer rotation was already applied if you loaded a mesh sample shape using LoadSampleShape so it wasn't a v consistent start point.

Does this look OK to you? Seems generally a good idea but thought I'd check in case you're managing this some other way. It will affect absorption corrections run on a workspace with a goniometer rotation.

gvardany commented:
We do use SetSample, but I'm not aware of interacting with goniometer directly. I assume if we just set a flat plate at an angle with SetSample, nothing should change right?

DannyHindson commented:
> We do use SetSample, but I'm not aware of interacting with goniometer directly. I assume if we just set a flat plate at an angle with SetSample, nothing should change right?

If the workspace doesn't have any goniometer entries in the sample logs then nothing should change. Even if the workspace does have some goniometer rotations in the sample logs then you usually have to explicitly activate them by using the SetGoniometer algorithm - there are one or two exceptions in that some of the LoadXXX algorithms call SetGoniometer internally (eg LoadEventNexus, LoadNXSPE) but if you're not using those you shouldn't have an active goniometer and nothing will change

peterfpeterson commented:
We are currently assuming all samples a spheres so it shouldn't be an issue

DanielMurphy22 committed:
New method ShapeFactory::addGoniometerTag()

Remove unnecessary virtual .getShapeXML()
Find </samplegeometry> tag to place goniometer tag
No longer look for </algebra> tag

DannyHindson commented on Framework/DataHandling/src/SetSample.cpp:
```diff
@@ -711,8 +712,10 @@ void SetSample::setSampleShape(API::ExperimentInfo &experiment, const Kernel::Pr
-    const auto xml = tryCreateXMLFromArgsOnly(*args, *refFrame);
+    auto xml = tryCreateXMLFromArgsOnly(*args, *refFrame);
     if (!xml.empty()) {
+      std::vector<double> rotationMatrix = experiment.run().getGoniometer().getR();
```
I think we should try and identify here whether the workspace has a goniometer set up with a non-default rotation matrix. It might have been helpful if getGoniometer returned nothing if this was the case, however it seems to be the case that the Run class create a default Goniometer object in its constructor.
Another way might be to see if the rotationMatrix returned is equal to an "identity" matrix or null rotation. We could then only insert the goniometer tag if it's a non-null rotation. This makes the code change a bit less intrusive in cases where an instrument doesn't have a goniometer.

DannyHindson commented on Framework/Geometry/src/Objects/ShapeFactory.cpp:
```diff
@@ -1416,5 +1607,92 @@ void ShapeFactory::createGeometryHandler(Poco::XML::Element *pElem, std::shared_
+  // Put goniometer tag in correct place in xml
+  std::size_t gonioPlace;
+  std::size_t foundType = xml.find("</type>");
+  std::size_t foundSampleGeometry = xml.find("</samplegeometry");
```
Having thought about it, I'm not so sure we should be adding the goniometer tag in situations where the CSG shape info is wrapped in a samplegeometry tag. This happens when SetSample is called with both the Environment and the Geometry parameters populated. In this situation a container geometry and a sample geometry is read in from a sample environment .xml file and just the sample part gets overridden by the value in the Geometry parameter. I think the use case here is that the user is saying that the can is only half full of powder so they load the can and sample shapes in from the .xml file but the sample shape needs updating.

I think at the moment we're only rotating the sample shape. This opens up a potential inconsistency where the sample shape is rotated but the can isn't. In reality this is unlikely to happen I think because goniometers are normally used on single crystal instruments where there isn't a container. So rather than try to handle this, it might be safer just not to insert the goniometer tag in sample shape information read in from a sample environment file

DanielMurphy22 replied:
Do I identify that the sample shape information is read in from a sample environment file by searching for the samplegeometry tag?

DannyHindson replied:
Yes that's one way of doing it. If you put the extra condition inside SetSample::setSampleShape then you could also check whether the sampleEnv parameter is empty or not. Perhaps this is more simple. Something like this:
```
  if (args) {
    const auto refFrame = experiment.getInstrument()->getReferenceFrame();
    auto xml = tryCreateXMLFromArgsOnly(*args, *refFrame);
    if (!xml.empty()) {
      std::vector<double> rotationMatrix = experiment.run().getGoniometer().getR();
      if (!sampleEnv)
        xml = Geometry::ShapeFactory().addGoniometerTag(rotationMatrix, xml);
      CreateSampleShape::setSampleShape(experiment, xml);
      return;
    }
  }
```

DanielMurphy22 replied:
I think my latest commit deals with both of these issues. I think we were actually safe anyway with regards to the sampleenv as in the case I tried, when `sampleEnv` was true, `args` was false.

DannyHindson replied:
I think the SNS system test that was failing a couple of days ago calls SetSample with both the Environment and Geometry parameters populated (which corresponds to sampleEnv=true and args also being true). This is a bit of a special case. Normally you'd just call SetSample with the Environment parameter populated which is what we've been doing in our conversations earlier today.
Just want to double check the new if statement - should the two conditions be combined with && ie both need to be true for you to add the goniometer tag?

DanielMurphy22 replied:
I've changed it between OR and AND a few times, but I think you're right it should be AND.

DannyHindson reviewed (commented):
Just had another look and have another couple of suggestions. Since this is a bit complicated we may need to have a chat. Just going for lunch but will be back in about an hour

DanielMurphy22 committed:
Only add goniometer tag if:

RotationMatrix is not the Identity Matrix
This shape is not defined within a SampleEnvironment

DanielMurphy22 committed:
Expose Container and Rotate Shapes together

DanielMurphy22 committed:
Add Rotate-all tag

New rotate-all tag allows the individual shapes in a
merged sample shape to move together by also rotating their centres
The individual rotate tag does NOT move a shapes centre

DannyHindson commented on Framework/Geometry/src/Objects/ShapeFactory.cpp:
```diff
@@ -963,6 +1145,43 @@ Hexahedron ShapeFactory::parseHexahedron(Poco::XML::Element *pElem) {
+    hex.rbb.rotate(m_gonioRotateMatrix);
+    hex.rbt.rotate(m_gonioRotateMatrix);
+  }
+
```
This is quite repetitive. I think it's worth adding a rotate method to the Hexahedron struct that rotates all 8 fields and then you can just call that method here. Although it's not essential, it's probably good "style" to change the struct to a class also

DannyHindson commented on Framework/Geometry/src/Objects/ShapeFactory.cpp:
```diff
@@ -1045,6 +1265,45 @@ std::string ShapeFactory::parseTaperedGuide(Poco::XML::Element *pElem, std::map<
+    hex.rbt.rotate(m_gonioRotateMatrix);
+    centre.rotate(m_gonioRotateMatrix);
+  }
+
```
This would also benefit from a rotate method on the Hexahedron struct

DannyHindson commented on Framework/Geometry/src/Objects/ShapeFactory.cpp:
```diff
@@ -1416,5 +1709,92 @@ void ShapeFactory::createGeometryHandler(Poco::XML::Element *pElem, std::shared_
+  const double cosZ = cos(zrotate);
+  std::vector<double> matrixList = {cosZ, -sinZ, 0, sinZ, cosZ, 0, 0, 0, 1};
+  return Kernel::Matrix<double>(matrixList);
+}
```
The 4 methods above (generateMatrix, generateXRotation, generateYRotation, generateZRotation) feel like they should be done using some pre-existing Mantid code rather than writing your own methods to achieve this. The way Mantid manages rotations is using the Quat class (short for Quaternion). I think you can rewrite generateMatrix to use the Quat class instead.

Unfortunately the Quaternion concept is a bit dense so you may need a lie down first! Fortunately the Goniometer::recalculateR method is some code which is pretty similar to what you want. It loops through the goniometer axes creating a Quat for each and then multiplies them together. Your code needs to do something similar by looping through the 3 x, y, z axes

It's probably worth getting the unit tests to where you want them to be before doing this change so you can be sure it doesn't break anything.

DanielMurphy22 replied:
I did conveniently ignore the Quat rotation as it seems quite confusing, but happy to have a go (as you rightly said) after I get some unit tests working. I did get these 4 methods from Mantid, I copied them across from MeshFileIO.cpp.  Otherwise I would have had to give Framework/Geometry access to all of Framework/DataHandling to avoid this copy.

DannyHindson commented on Framework/Geometry/src/Objects/ShapeFactory.cpp:
```diff
@@ -614,32 +766,61 @@ CuboidCorners ShapeFactory::parseCuboid(Poco::XML::Element *pElem) {
+      result.lfb.rotate(m_gonioRotateMatrix);
+      result.lft.rotate(m_gonioRotateMatrix);
+      result.lbb.rotate(m_gonioRotateMatrix);
+      result.rfb.rotate(m_gonioRotateMatrix);
```
Similar comment to the one I've made elsewhere about the Hexahedron. This code to rotate the four corners could perhaps benefit from being put into a method on the CuboidCorners struct

DannyHindson commented on Framework/DataHandling/src/SetSample.cpp:
```diff
@@ -711,8 +712,14 @@ void SetSample::setSampleShape(API::ExperimentInfo &experiment, const Kernel::Pr
+    auto xml = tryCreateXMLFromArgsOnly(*args, *refFrame);
     if (!xml.empty()) {
+      std::vector<double> rotationMatrix = experiment.run().getGoniometer().getR();
+      if (rotationMatrix != std::vector<double>{1, 0, 0, 0, 1, 0, 0, 0, 1} && !sampleEnv) {
```
The Matrix class provides a method to return an identity matrix `Matrix<T>::identityMatrix()`. Please can you create a matrix using that to compare against? You could just do `rotationMatrix.identityMatrix()` I think

DanielMurphy22 replied:
The rotationMatrix isn't of type Matrix, it's a std::vector of the 9 inputs so I'm not sure I can call .identityMatrix or compare it to rotationMatrix. Do you think it's worth changing all rotationMatrix's to type Matrix?

DannyHindson reviewed (changes requested):
I've not tested CopySample yet but based on my tests of the 3 new rotation tags using SetSample it all looks good so far. I've added a few detail level comments. A couple of general ones:

- please can you update the CSG documentation to add a description of the new rotation tags (HowToDefineGeometricShape.rst) + include comments on the shapes that don't support it eg cone

- I had a look through the code where you rotate the corners of the shape and also separately rotate the centre (conversation we had earlier). I think this is equivalent to offsetting the corners to the centre and then rotating the corners. I think this is an example of the distributive property of matrix multiplication 
ie if V is a vector describing a corner of the shape and C is a vector pointing at the centre:
R(V+C) = RV +RC
So if you first offset the corners to the centre and then rotate (lhs) you get the same answer as rotating the corners and centre separately and then adding together (rhs)

DannyHindson commented on Framework/DataHandling/test/SetSampleTest.h:
```diff
@@ -1135,4 +1135,4 @@ class SetSampleTest : public CxxTest::TestSuite {
   const std::string m_instName = "ISIS_Histogram";
   const std::string m_envName = "TestEnv";
-};
+};
```
Not sure what the GitHub diff is seeing here but can we remove this difference? Perhaps there's a carriage return missing in the new version

DanielMurphy22 committed:
ShapeFactory Rotate tests

Move GenerateRotationMatrix from MeshFileIO to ShapeFactory
Add tests for rotating cuboid, cylinder, hollow cylinder, slice,
tapered-guide, sphere, and infinite-plane/cylinder
Also rotate by correct matrix
Fix addGoniometertag

DanielMurphy22 commented:
This xml currently doesn't work: merge_xml = 
``` xml
' \
<cylinder id="stick"> \
  <centre-of-bottom-base x="-1.0" y="0.0" z="0.0" /> \
  <axis x="1.0" y="0.0" z="0.0" /> \
  <radius val="0.2" /> \
  <height val="2.0" /> \
</cylinder> \
 \
<sphere id="some-sphere"> \
  <centre x="1.5"  y="0.0" z="0.0" /> \
  <radius val="0.5" /> \
</sphere> \
\
<algebra val="some-sphere (: stick)" /> \
<rotate-all x="0" y="90" z="0" /> \
'
```

DanielMurphy22 commented on Framework/DataHandling/src/MeshFileIO.cpp:
```diff
@@ -5,26 +5,11 @@
 //   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
 // SPDX - License - Identifier: GPL - 3.0 +
 #include "MantidDataHandling/MeshFileIO.h"
+#include "MantidGeometry/Objects/ShapeFactory.h"
```
I think the cppcheck failure is due to this although this code does work

DanielMurphy22 replied:
https://builds.mantidproject.org/job/pull_requests-cppcheck/47570/cppcheckResult/source.52/

DannyHindson commented on Framework/Algorithms/inc/MantidAlgorithms/CopySample.h:
```diff
@@ -55,7 +55,8 @@ class MANTID_ALGORITHMS_DLL CopySample : public API::Algorithm {
   void copyParameters(API::Sample &from, API::Sample &to, bool nameFlag, bool materialFlag, bool environmentFlag,
-                      bool shapeFlag, bool latticeFlag, bool orientationOnlyFlag);
+                      bool shapeFlag, bool latticeFlag, bool orientationOnlyFlag,
+                      Kernel::Matrix<double> rotationMatrix);
```
Can rotationMatrix by passed by const reference?

DannyHindson commented on Framework/Algorithms/src/CopySample.cpp:
```diff
@@ -111,7 +115,8 @@ void CopySample::exec() {
         copyParameters(sample, outMDWS->getExperimentInfo(i)->mutableSample(), copyName, copyMaterial, copyEnvironment,
-                       copyShape, copyLattice, copyOrientation);
+                       copyShape, copyLattice, copyOrientation,
+                       outMDWS->getExperimentInfo(i)->run().getGoniometer().getR());
```
You could do `auto ei = outMDWS->getExperimentInfo(i)` before the call to copyParameters just to make the argument list is a bit less verbose

DannyHindson commented on Framework/Algorithms/src/CopySample.cpp:
```diff
@@ -120,22 +125,25 @@ void CopySample::exec() {
+      copyParameters(
+          sample, outMDWS->getExperimentInfo(static_cast<uint16_t>(outputSampleNumber))->mutableSample(), copyName,
+          copyMaterial, copyEnvironment, copyShape, copyLattice, copyOrientation,
+          outMDWS->getExperimentInfo(static_cast<uint16_t>(outputSampleNumber))->run().getGoniometer().getR());
```
Similar to above, I suggest you do `auto ei = outMDWS->getExperimentInfo(static_cast<uint16_t>(outputSampleNumber))` before the call to copyParameters

DannyHindson commented on Framework/Geometry/inc/MantidGeometry/Objects/ShapeFactory.h:
```diff
@@ -75,6 +94,11 @@ class MANTID_GEOMETRY_DLL ShapeFactory {
   static std::shared_ptr<CSGObject> createSphere(const Kernel::V3D &centre, double radius);
   static std::shared_ptr<CSGObject> createHexahedralShape(double xlb, double xlf, double xrf, double xrb, double ylb,
                                                           double ylf, double yrf, double yrb);
+  std::string addGoniometerTag(Kernel::Matrix<double> rotateMatrix, std::string xml);
```
Can rotateMatrix be passed by const ref?

DannyHindson commented on Framework/Geometry/inc/MantidGeometry/Objects/ShapeFactory.h:
```diff
@@ -83,8 +107,8 @@ class MANTID_GEOMETRY_DLL ShapeFactory {
-  std::string parseSegmentedCylinder(Poco::XML::Element *pElem, std::map<int, std::shared_ptr<Surface>> &prim,
-                                     int &l_id);
+  // std::string parseSegmentedCylinder(Poco::XML::Element *pElem, std::map<int, std::shared_ptr<Surface>> &prim,
+  //                                    int &l_id);
```
Please can you either delete these lines or if they're were temporarily commented out reinstate?

DanielMurphy22 replied:
This method has no references and the SegmentedCylinder entry in ShapeInfo::GeometryShape was removed after this convo: https://github.com/mantidproject/mantid/pull/21310#discussion_r154614470

So I'll delete the parseSegmentedCylinder() method

DannyHindson commented on Framework/DataHandling/src/SetSample.cpp:
```diff
@@ -31,6 +31,7 @@
 #include <Poco/Path.h>
 #include <boost/algorithm/string/case_conv.hpp>
 #include <boost/algorithm/string/predicate.hpp>
+#include <string>
```
Is this still needed? Seems to compile without on Windows at least

DannyHindson commented on Framework/Geometry/src/Objects/ShapeFactory.cpp:
```diff
@@ -53,6 +53,16 @@ const V3D DEFAULT_AXIS(0, 0, 1);
 } // namespace
 
+namespace {
+std::vector<double> DegreesToRadians(std::vector<double> anglesDegrees) {
```
Pass anglesDegrees as const ref

DannyHindson commented on docs/source/release/v6.2.0/framework.rst:
```diff
@@ -45,6 +45,8 @@ Fit Functions
 ------------
 - **Sample shapes which are CSGObjects can now be plotted. Shapes can also be merged, such as a sphere with a cylindrical hole. For more details see** :ref:`Mesh_Plots`.
+- Sample shapes (both MeshObjects and CSGObjects) can be manually rotated. Rotations from :ref:`SetGoniometer <algm-SetGoniometer>` are applied. This works with :ref:`CopySample <algm-CopySample>`,
+  so the copied shape can be plotted, but the old goniometer angle is not copied.
```
Please can you also mention that SetSample has changed - I'm thinking of the extra support for the user facing rotate and rotate-all tags

DannyHindson commented on Framework/DataHandling/src/SetSample.cpp:
```diff
@@ -711,8 +712,14 @@ void SetSample::setSampleShape(API::ExperimentInfo &experiment, const Kernel::Pr
     if (!xml.empty()) {
+      Kernel::Matrix<double> rotationMatrix = experiment.run().getGoniometer().getR();
+      if (rotationMatrix != Kernel::Matrix<double>(3, 3, 1) && !sampleEnv) {
+        // Only add goniometer tag if rotataionMatrix is not the Identity,
```
typo in rotataionMatrix

DannyHindson commented on Framework/Geometry/inc/MantidGeometry/Objects/ShapeFactory.h:
```diff
@@ -39,6 +40,13 @@ struct CuboidCorners {
   Kernel::V3D lbb;
   Kernel::V3D rfb;
+
+  void rotatePoints(std::vector<double> rotationMatrix) {
```
Please can you pass by const ref

DannyHindson commented on Framework/Geometry/inc/MantidGeometry/Objects/ShapeFactory.h:
```diff
@@ -50,6 +58,17 @@ struct Hexahedron {
   Kernel::V3D rbb; // right back bottom
   Kernel::V3D rbt; // right back top
+
+  void rotatePoints(std::vector<double> rotationMatrix) {
```
Please can you pass by const ref

DanielMurphy22 replied:
I sense a theme...

DannyHindson reviewed (changes requested):
Good job on this! This is quite a complicated piece of work but the code changes look nice and clean now. I've added a few detail level comments but nothing difficult\major. I've retested and all seems to work as I'd expect

Please can you update the test scripts on the PR as well to use the final name for the "rotate" tag. They still use the older "rotation" name. It might be useful to refer back to these in future

DanielMurphy22 committed:
Final detail changes

Pass by const reference where possible
Make code cleaner andremove unneccessary include
Update release notes
Add comments to CopySample and SetSample

DannyHindson reviewed (approved):
Thanks for the latest changes. All looks good now
